torch/distributed/_tensor/experimental/tensor_parallel.py

```diff
@@ -0,0 +1,426 @@
+import copy
+import operator
+from typing import Any, cast, Dict, List, Optional, Sequence, Set, Tuple
+
+import torch
+from torch.distributed._tensor import DeviceMesh, distribute_tensor, DTensor
+from torch.distributed._tensor.op_schema import (
+    DTensorSpec,
+    OpSchema,
+    OutputSharding,
+    OutputSpecType,
+    PlacementStrategy,
+)
+from torch.distributed._tensor.placement_types import (
+    Placement,
+    Replicate,
+    Shard,
+    TensorMeta,
+)
+from torch.distributed._tensor.redistribute import redistribute_local_tensor
+from torch.export import ExportedProgram
+from torch.export.exported_program import ExportGraphSignature
+from torch.fx import GraphModule
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.fx.node import Node
+from torch.fx.passes.infra.pass_base import PassBase, PassResult
+from torch.fx.passes.shape_prop import _extract_tensor_metadata
+
+
+aten = torch.ops.aten
+
+
+def tensor_parallel_transformation(
+    exported_program: ExportedProgram,
+    rank: int,
+    world_size: int,
+) -> ExportedProgram:
+    return exported_program._transform(
+        TensorParallelTransformPass(
+            rank,
+            world_size,
+            exported_program.state_dict,
+            exported_program.graph_signature,
+        )
+    )
+
+
+class TensorParallelTransformPass(PassBase):
+    def __init__(
+        self,
+        rank: int,
+        world_size: int,
+        state_dict: Dict[str, torch.Tensor],
+        graph_signiture: ExportGraphSignature,
+    ) -> None:
+        super().__init__()
+        self.rank = rank
+        self.mesh = DeviceMesh("cuda", torch.arange(world_size))
+        self.state_dict: Dict[str, torch.Tensor] = state_dict
+        self.graph_signiture = graph_signiture
+
+    def call(self, graph_module) -> PassResult:
+        gm = copy.deepcopy(graph_module)
+
+        placement_strategies = mark_sharding(gm, self.graph_signiture, self.mesh)
+        partitioner(gm)
+        shard_state_dict(
+            self.state_dict, placement_strategies, self.graph_signiture, self.mesh
+        )
+        return PassResult(gm, True)
+
+
+def mark_tensor_parallel_shardings(
```

Comment:
maybe we should made all those APIs be private (i.e. by prepanding `_`) as they are all exclusive to this file only. let's only leave a public `tensor_parallel_transformation` API and make the rest to be private